homeassistant/components/homeassistant/__init__.py

```diff
@@ -23,13 +23,17 @@
 from homeassistant.helpers import config_validation as cv
 from homeassistant.helpers.service import async_extract_referenced_entity_ids
 
+ATTR_CONFIG_ENTRY_ID = "config_entry_id"
+
 _LOGGER = logging.getLogger(__name__)
 DOMAIN = ha.DOMAIN
 SERVICE_RELOAD_CORE_CONFIG = "reload_core_config"
+SERVICE_RELOAD_CONFIG_ENTRY = "reload_config_entry"
 SERVICE_CHECK_CONFIG = "check_config"
 SERVICE_UPDATE_ENTITY = "update_entity"
 SERVICE_SET_LOCATION = "set_location"
 SCHEMA_UPDATE_ENTITY = vol.Schema({ATTR_ENTITY_ID: cv.entity_ids})
+SCHEMA_RELOAD_CONFIG_ENTRY = vol.Schema({vol.Required(ATTR_CONFIG_ENTRY_ID): str})
```

Comment:
Instead of entry_id, maybe require either integration domain or entity_id. If no entity_id is passed all config entries for the domain are reloaded. If an entity_id is passed we look up the config entry from the entity registry.

Reply:
<img width="569" alt="Screen Shot 2021-03-10 at 9 07 46 PM" src="https://user-images.githubusercontent.com/663432/110749085-b5f21580-81e4-11eb-86f4-6ccc151d0df9.png">

I've made it accept a target, but I can't seem to find a way to make the target optional since entries like `homekit` will have no entities.

I guess I could make two services but I'd rather find a way to make it optional.

Reply:
Config entry entry_id is not user facing. I don't think we should accept that at all.

Reply:
I can take it out, but it doesn't solve for the original case where we want to reload a specific homekit config entry. Reloading them all can be quite expensive.

Reply:
HomeKit has a virtual bridge device so they can hit that. I'll take the rest out.

Reply:
Would it be possible to lookup the correct aid storage via an entity_id? It looks like each storage is mapped to a single config entry.

Reply:
I can lookup the config entry id from the device since `async_extract_entity_ids` is not going to have any entities to match in this case.  Will give this another pass when I have another free window.

Reply:
I think we need a `async_extract_referenced_config_entry_ids` to match `async_extract_referenced_entity_ids` because by the time we convert to entity ids we are assuming the integration provides entities.